src/SLCore/Common/Models/FileUri.cs

```diff
@@ -28,15 +30,38 @@ namespace SonarLint.VisualStudio.SLCore.Common.Models;
 public sealed class FileUri
 {
     private readonly Uri uri;
+    private static readonly char[] Rfc3986ReservedCharsToEncoding = ['?', '#', '[', ']', '@'];
 
     public FileUri(string uriString)
     {
-        uri = new Uri(uriString);
+        var unescapedUri = Uri.UnescapeDataString(uriString);
+        uri = new Uri(unescapedUri);
     }
 
     public string LocalPath => uri.LocalPath;
 
-    public override string ToString() => Uri.EscapeUriString(uri.ToString());
+    public override string ToString()
+    {
+        var escapedUri = Uri.EscapeUriString(uri.ToString());
+
+        return EscapeRfc3986ReservedCharacters(escapedUri);
+    }
+
+    /// <summary>
+    /// The backend (SlCore) uses java, in which the Uri follows the RFC 3986 protocol.
+    /// The <see cref="Uri.EscapeUriString"/> does not escape the reserved characters, that's why they are escaped here.
+    /// See https://learn.microsoft.com/en-us/dotnet/api/system.uri.escapeuristring?view=netframework-4.7.2
+    /// </summary>
+    /// <param name="stringToEscape"></param>
+    /// <returns></returns>
+    private static string EscapeRfc3986ReservedCharacters(string stringToEscape)
+    {
+        var charsToEscape = Rfc3986ReservedCharsToEncoding.Where(stringToEscape.Contains).ToList();
+
+        return !charsToEscape.Any()
+            ? stringToEscape
+            : charsToEscape.Aggregate(stringToEscape, (current, charToEscape) => current.Replace(charToEscape.ToString(), Uri.HexEscape(charToEscape)));
+    }
```

Comment:
Yes, line 59 gives you O(n) complexity, which will probably be 99% of the cases, because I don't expected that the Rfc3986ReservedChars to exist in file names too often.

Reply:
That's the thing, you could just skip that line and leave the rest of the code as is (charToEscape would be equal to all reserved characters), and it would actually reduce the number of iterations for the same input. Using a string builder is optional. If you think that more often than not it's going to be returning the same string, then we can use string replace aggregate to save on the StringBuilder creation. Though, it takes nanoseconds for `new StringBuilder(string).ToString()` to execute